### PR TITLE
identity/redisdir: various fixes

### PR DESCRIPTION
**Pull request by bnewbold**
A bunch of cleanups and bug fixes.

The biggest fix is to apply a bug fix from the cache directory resulting in "nil/nil" responses. These were fixed in https://github.com/bluesky-social/indigo/pull/479, but not copied to the redis directory implementation, which has the same coalescing code. This recently impacted @whyrusleeping's use in discover. This is a backwards-compatible fix.

Another fix handles cached error messages. An empty string was being stored for the DID. Because redis cache serializes, and the de-serialized DID is re-parsed, the empty strings fail to parse, resulting in an error. I guess we could consider having text parsing of DIDs not actually parse the string? but that would break auto-validation when it is expected. The fix was to make that field nullable. this might make the change not-backwards-compatible with existing cached metadata? this only impacts cached errors, which have a shorter TTL, so might not be that big of a deal in practice?

Other smaller cleanups and fixes, which rolled in here to make the code align better with the in-memory caching directory, so that merge/review is easier:

- adds invalid handle TTL to redisdir
- adds "WithCacheState" variants of methods, for metrics/perf measurement
- fix deletion/purging
- special-case handling of resolving invalid handle

One remaining issue is that the cached error types come back as generic errors (with error kind/type striped). It would be nice if that wasn't the case, so that code could detect specific types of errors? Considering doing a manual enum/integer hack to encode the type and convert the errors to the actual type on read, for a small set of known identity errors.

### CODE REVIEW

**Review by brianolson — APPROVED** [review]
I think I get it, mostly improving error handling. same quibble that I prefer `errors.New("")` when the string doesn't need formatting params

File: atproto/identity/redisdir/redis_directory.go
Comment (whyrusleeping) [inline atproto/identity/redisdir/redis_directory.go:168]: `err == nil`

File: atproto/identity/redisdir/redis_directory.go
Comment (whyrusleeping) [inline atproto/identity/redisdir/redis_directory.go:300]: `err == nil`

File: atproto/identity/redisdir/redis_directory.go
Comment (whyrusleeping) [inline atproto/identity/redisdir/redis_directory.go:366]: why is this backwards everywhere? is this a rust thing?

Comment (ericvolp12) [reply atproto/identity/redisdir/redis_directory.go:366]: I think Bryan does this intentionally to make it clear it's a `==` and not a `err != nil`

**Comment by bnewbold** [comment]
Resolved the review things. My disposition is to merge this, but I want to monitor when this gets deployed (in various places), so i'll probably wait until i'm ready to do a controlled deploy and merge it then.

**Pull request merged**
